### ISSUE TITLE
Validator and disassembler produce vague errors

### ISSUE THREAD

**Issue by Bidski** [KhronosGroup/SPIRV-Tools#2091]
I just installed the master branch so that I can validate a spirv module, but I am getting a very vague error. 

I was under the impression that there was functionality to display the offending instruction in the validator, but this is not happening with the master branch? Did this get removed?

```
$ spirv-val ./fetch.spv
error: line 122: Id is 0
```

The disassembler also doesn't give much insight. I **think** that the disassembler has stopped (just before?) the offending instruction, but from what I can see of the instructions that should follow none of them have an ID of 0.
```
$ spirv-dis --raw-id --offsets ./fetch.spv
; SPIR-V
; Version: 1.3
; Generator: Khronos; 0
; Bound: 144
; Schema: 0
               OpCapability Addresses ; 0x00000014
               OpCapability Linkage ; 0x0000001c
               OpCapability Kernel ; 0x00000024
               OpCapability ImageBasic ; 0x0000002c
               OpCapability LiteralSampler ; 0x00000034
          %1 = OpExtInstImport "OpenCL.std" ; 0x0000003c
               OpMemoryModel Physical32 OpenCL ; 0x00000050
          %2 = OpTypeInt 32 0 ; 0x0000005c
          %3 = OpTypeFloat 32 ; 0x0000006c
          %4 = OpTypeBool ; 0x00000078
          %5 = OpTypeVector %2 2 ; 0x00000080
          %6 = OpTypeVector %2 3 ; 0x00000090
          %7 = OpTypeVector %2 4 ; 0x000000a0
          %8 = OpTypeVector %3 2 ; 0x000000b0
          %9 = OpTypeVector %3 3 ; 0x000000c0
         %10 = OpTypeVector %3 4 ; 0x000000d0
         %15 = OpTypeImage %3 2D 0 0 0 1 Unknown ReadOnly ; 0x000000e0
         %16 = OpTypeSampler ; 0x00000108
         %17 = OpTypePointer UniformConstant %16 ; 0x00000110
         %18 = OpTypeSampledImage %15 ; 0x00000120
         %19 = OpTypeFunction %3 %15 %16 %8 ; 0x0000012c
         %28 = OpTypeFunction %10 %15 %16 %8 %8 ; 0x00000144
         %11 = OpConstant %2 0 ; 0x00000160
         %12 = OpConstant %3 0 ; 0x00000170
         %13 = OpConstant %2 1 ; 0x00000180
         %14 = OpConstant %3 1 ; 0x00000190
         %41 = OpConstantComposite %10 %3221225472 %3212836864 %1065353216 %1073741824 ; 0x000001a0
         %42 = OpConstantComposite %10 %3187671040 %3191865344 %1031798784 %3187671040 ; 0x000001bc
error: 122: Id is 0
```

How can I get more information about this error?

**Comment by Bidski** [comment]
I figured out the issue (mainly me not reading the documentation properly).

The issue is that `OpConstantComposite` wants the IDs of `OpConstant`s which it uses to construct the composite and not the literal values. I just blindly assumed that `OpConstant` and `OpConstantComposite` had the same structure to them.

Regardless of this, I feel that the reported error is very misleading in this case.

**Comment by alan-baker** [comment]
It would be nice if the validator gave an error that `%3221225472` was not defined.

**Comment by dj2** [comment]
Ah, I see, so specifically,

```
%41 = OpConstantComposite %10 %3221225472 %3212836864 %1065353216 %1073741824
%42 = OpConstantComposite %10 %3187671040 %3191865344 %1031798784 %3187671040
```

those two lines are the problem because the values don't map to existing ids, yea?

**Comment by alan-baker** [comment]
Yes, I think so. It may be a matter of the order of checks, but you'll have to look into it.

**Comment by dj2** [comment]
With:

```
OpCapability Shader
OpCapability Float64
OpMemoryModel Logical GLSL450
OpEntryPoint Fragment %main "main"
OpExecutionMode %main OriginUpperLeft
%void = OpTypeVoid
%func = OpTypeFunction %void

%f32 = OpTypeFloat 32
%f32vec4 = OpTypeVector %f32 4

%main = OpFunction %void None %func
%val1 = OpCompositeConstruct %f32vec4 %0 %222 %333 %444

OpReturn
OpFunctionEnd
```

I get:
```
error: 11: ID 7 has not been defined
  %6 = OpCompositeConstruct %v4float %7 %8 %9 %10
```
Which does have a disassembly. I think this output is fine, we re-mapped the id's but there is a flag to not have it do that.


If I change the %0 to just 0 I get:
```
ERROR: Expected id to start with %.
```
Which is really un-helpful as it has no disassembly and no line number.

@Bidski did you do the %0 or 0 in your shader which caused the error?

**Comment by dj2** [comment]
Ah the ERROR line comes from the unit test frame work and it's because we can't turn the text into a binary.

**Comment by dneto0** [comment]
The disassembly of the offending instruction is only available when processing actually gets into the validator.
But if the instructions are malformed so that the binary parse fails, then you won't get that far.  So really it's an issue of spvBinaryParse.    We can't use the regular print-the-bad-instruction facility because we don't actually *have* an instruction in hand.

I grant you 100% that "Id is 0" is a bad error message.  Should have been something like "error: <instruction number>: Expected an Id operand but got undefined Id %<blah>".

**Comment by Bidski** [comment]
> Can you share a spirv-asm file which shows the error so we can try to get better messaging here?

Unfortunately I neglected to save the offending file.

> `%41 = OpConstantComposite %10 %3221225472 %3212836864 %1065353216 %1073741824`
> `%42 = OpConstantComposite %10 %3187671040 %3191865344 %1031798784 %3187671040`

These two lines would be the issue (along with other similar lines). The numbers there are actually floating point literals represented as ints, which are obviously being interpreted as IDs and not literals (this is my mistake for not reading the docs properly).

> @Bidski did you do the %0 or 0 in your shader which caused the error?

I did neither of these. I am actually writing the raw words to a binary file and not writing the assembly code. What you see in the `OpConstantComposite` instructions above is literally what I wrote to the file in binary form.

For reference, I am not using a GLSL style shader, but an OpenCL style kernel.
